File: src-tauri/src/coordinator.rs
use std::{collections::HashMap, sync::Mutex, time::Instant};
// `State` and `WebviewWindow` come back in A6 with the commands; importing
// them now would be an unused_imports warning across A1-A5.
use tauri::{AppHandle, Emitter};

/// One PTY event held back while its pane is mid-transfer.
#[derive(Clone, Debug, PartialEq)]
pub struct BufferedEvent {
    pub event: String,
    pub payload: serde_json::Value,
}

/// Where a pane's output goes right now.
enum PaneRoute {
    /// Steady state — delivered to this window label.
    Owned(String),
    /// A transfer is open; output is buffered, not delivered (§7.2).
    Transferring(Transfer),
}

/// An open transfer. The spec sketches this as inline enum fields; a named
/// struct keeps every lookup helper returning `&mut Transfer` instead of
/// re-destructuring the variant at each call site.
struct Transfer {
    /// Window the pane is leaving.
    from: String,
    /// Window that claimed it. `None` until `claim_transfer`.
    to: Option<String>,
    /// Window a pending adoption was registered for, before it claims. Lets a
    /// destination that dies before `claim` still abort the transfer (§7.6).
    reserved_to: Option<String>,
    token: String,
    buffered: Vec<BufferedEvent>,
    buffered_bytes: usize,
    /// The PTY exited mid-transfer. The route entry must outlive that so the
    /// buffered exit event still reaches the destination (§7.6).
    exited: bool,
    started: Instant,
}

/// Where the coordinator emits. Production is `AppSink`; tests record, which is
/// the only way to assert delivery order — `AppHandle` cannot be constructed in
/// a unit test.
pub trait EventSink {
    fn emit(&self, label: &str, event: &str, payload: &serde_json::Value);
}

pub struct AppSink<'a>(pub &'a AppHandle);

impl EventSink for AppSink<'_> {
    fn emit(&self, label: &str, event: &str, payload: &serde_json::Value) {
        let _ = self.0.emit_to(label.to_string(), event, payload);
    }
}

/// Everything the coordinator guards, behind ONE mutex. Routes and settled
/// tokens must move together: a commit changes both, and a reader that saw one
/// without the other would answer wrongly.
#[derive(Default)]
struct CoordinatorState {
    routes: HashMap<u32, PaneRoute>,
}

/// App-level pane → window routing. Routes PTY output/exit to the owning
/// webview only, and holds output still across a pane transfer (§7).
#[derive(Default)]
pub struct WindowCoordinator {
    state: Mutex<CoordinatorState>,
}

impl WindowCoordinator {
    pub fn register(&self, pane_id: u32, window_label: String) {
        if let Ok(mut state) = self.state.lock() {
            state.routes.insert(pane_id, PaneRoute::Owned(window_label));
        }
    }

    pub fn unregister(&self, pane_id: u32) {
        if let Ok(mut state) = self.state.lock() {
            state.routes.remove(&pane_id);
        }
    }

    /// The owning window, or `None` while a transfer is open — a transferring
    /// pane has no owner, and saying otherwise would let a caller act on it.
    #[allow(dead_code)] // window lifecycle census (§9.4)
    pub fn owner(&self, pane_id: u32) -> Option<String> {
        let state = self.state.lock().ok()?;
        match state.routes.get(&pane_id) {
            Some(PaneRoute::Owned(label)) => Some(label.clone()),
            Some(PaneRoute::Transferring(_)) | None => None,
        }
    }

    /// Pane ids still owned by this window (for close-window dispose).
    #[allow(dead_code)] // used when multi-window close lands
    pub fn panes_for_window(&self, window_label: &str) -> Vec<u32> {
        let Ok(state) = self.state.lock() else {
            return Vec::new();
        };
        state
            .routes
            .iter()
            .filter_map(|(id, route)| match route {
                PaneRoute::Owned(label) if label == window_label => Some(*id),
                _ => None,
            })
            .collect()
    }

    /// Route one PTY event under a SINGLE lock (§7.2). Emission happens while
    /// the lock is held, so a chunk read during a commit cannot overtake the
    /// flush that commit performs.
    pub fn deliver(
        &self,
        sink: &dyn EventSink,
        pane_id: u32,
        event: &str,
        payload: serde_json::Value,
        _now: Instant,
    ) {
        let Ok(mut state) = self.state.lock() else {
            eprintln!("Deck: route lock poisoned, dropping {event} for pane {pane_id}");
            return;
        };
        match state.routes.get_mut(&pane_id) {
            Some(PaneRoute::Owned(label)) => sink.emit(label, event, &payload),
            Some(PaneRoute::Transferring(transfer)) => {
                transfer.buffered.push(BufferedEvent {
                    event: event.to_string(),
                    payload,
                });
            }
            // No broadcast fallback: sending one window's terminal output to
            // every window is a data leak, not a safety net (§7.2).
            None => eprintln!("Deck: no route for pane {pane_id}, dropping {event}"),
        }
    }
}

/// Emit a PTY event to the pane's current route.
pub fn emit_to_owner<S: serde::Serialize + Clone>(
    app: &AppHandle,
    coordinator: &WindowCoordinator,
    pane_id: u32,
    event: &str,
    payload: S,
) {
    let value = match serde_json::to_value(payload) {
        Ok(value) => value,
        Err(error) => {
            eprintln!("Deck: cannot serialize {event} for pane {pane_id}: {error}");
            return;
        }
    };
    coordinator.deliver(&AppSink(app), pane_id, event, value, Instant::now());
}

/// Test double shared across the crate.
///
/// **This is `pub(crate)` on purpose — do not "clean up" an apparently unused
/// visibility.** The window-lifecycle tests live in other modules and drive the
/// coordinator through `&dyn EventSink` as well (their census test passes one
/// to `begin_transfer`), so a private double here would force a duplicate over
/// there. The path other modules import is
/// `crate::coordinator::test_support::RecordingSink`.
#[cfg(test)]
pub(crate) mod test_support {
    use super::EventSink;
    use std::sync::Mutex;

    /// Records what the coordinator emitted, in emission order. Stands in for
    /// `AppSink`, which needs an `AppHandle` no unit test can build.
    #[derive(Default)]
    pub(crate) struct RecordingSink {
        emitted: Mutex<Vec<(String, String, serde_json::Value)>>,
    }

    impl EventSink for RecordingSink {
        fn emit(&self, label: &str, event: &str, payload: &serde_json::Value) {
            self.emitted.lock().expect("recording sink lock").push((
                label.to_string(),
                event.to_string(),
                payload.clone(),
            ));
        }
    }

    impl RecordingSink {
        /// `(label, data)` for PTY output only, in order. Filtered on purpose:
        /// `transfer:settled` shares this stream from Task A3 onward, and an
        /// unfiltered view would make every ordering assertion here depend on
        /// a payload shape it is not testing.
        pub(crate) fn delivered(&self) -> Vec<(String, String)> {
            self.emitted
                .lock()
                .expect("recording sink lock")
                .iter()
                .filter(|(_, event, _)| event == "pty:output")
                .map(|(label, _, payload)| {
                    (
                        label.clone(),
                        payload["data"].as_str().unwrap_or_default().to_string(),
                    )
                })
                .collect()
        }

        /// `(label, event)` for every emission, in order — the unfiltered view.
        pub(crate) fn events(&self) -> Vec<(String, String)> {
            self.emitted
                .lock()
                .expect("recording sink lock")
                .iter()
                .map(|(label, event, _)| (label.clone(), event.clone()))
                .collect()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::test_support::RecordingSink;
    use super::WindowCoordinator;
    use std::time::Instant;

    /// Local to this module: the shared double is `RecordingSink`, not the
    /// payload builders. Another module testing its own concern builds its own.
    fn output(data: &str) -> serde_json::Value {
        serde_json::json!({ "id": 1, "data": data })
    }

    #[test]
    fn owned_route_delivers_to_the_owning_window_only() {
        let coordinator = WindowCoordinator::default();
        let sink = RecordingSink::default();
        coordinator.register(1, "main".into());
        coordinator.register(2, "deck-1".into());

        coordinator.deliver(&sink, 1, "pty:output", output("hello"), Instant::now());

        assert_eq!(
            sink.delivered(),
            vec![("main".to_string(), "hello".to_string())]
        );
        assert_eq!(coordinator.owner(1).as_deref(), Some("main"));
    }

    #[test]
    fn an_unrouted_pane_drops_the_chunk_instead_of_broadcasting() {
        let coordinator = WindowCoordinator::default();
        let sink = RecordingSink::default();

        coordinator.deliver(&sink, 7, "pty:output", output("secret"), Instant::now());

        assert!(
            sink.delivered().is_empty(),
            "output for an unrouted pane must be dropped, never broadcast"
        );
    }

    #[test]
    fn unregister_clears_an_owned_route() {
        let coordinator = WindowCoordinator::default();
        coordinator.register(1, "main".into());
        coordinator.unregister(1);
        assert_eq!(coordinator.owner(1), None);
    }

    #[test]
    fn panes_for_window_filters() {
        let coordinator = WindowCoordinator::default();
        coordinator.register(1, "a".into());
        coordinator.register(2, "b".into());
        coordinator.register(3, "a".into());
        let mut panes = coordinator.panes_for_window("a");
        panes.sort();
        assert_eq!(panes, vec![1, 3]);
    }
}

File: electron/coordinator.ts
/**
 * Pane → window routing and the pane-transfer transaction — the port of
 * `src-tauri/src/coordinator.rs`.
 *
 * This is a load-bearing seam. Eight blockers were found and fixed in the Rust
 * original by an adversarial review, and every one of them is a rule below
 * rather than an implementation detail. The invariants worth naming:
 *
 *  - **Output is never broadcast.** A pane with no route drops its event and
 *    says so. Sending one window's terminal output to every window is a data
 *    leak, not a safety net.
 *  - **A transferring pane has no owner.** Not the source, not the
 *    destination — for the duration of the handoff nobody may write to it.
 *  - **Flush before announce.** Buffered events go out before
 *    `transfer:settled`, so a destination has every byte before it learns the
 *    pane is its own.
 *  - **Tokens make retries idempotent.** A settled token keeps answering after
 *    the fact, so a duplicate commit or abort is not an error.
 *  - **Death rules are asymmetric.** A dead DESTINATION aborts the transfer; a
 *    dead SOURCE does not, because the destination can still claim and commit.
 *  - **A pane is never handed to a dead window.** It is killed instead —
 *    losing a pane is recoverable, leaking a PTY nobody reads is not.
 *
 * The Rust version holds one mutex over all of this. The main process is
 * single-threaded, so the lock is gone; what remains is the ordering, which is
 * preserved call for call.
 */

/** A transfer that has not committed by this point is abandoned back to its
 * source. Enforced lazily on every entry point rather than by a timer, so
 * there is no wakeup to schedule and no timer to leak. */
export const TRANSFER_TIMEOUT_MS = 10_000;

/** Ceiling on what one transfer may hold back. Past it the move is abandoned
 * and everything buffered goes to the source: losing the move is recoverable,
 * losing output is not. */
export const BUFFER_MAX_BYTES = 4 * 1024 * 1024;

/** How many finished tokens stay answerable. A retry arrives within one
 * transfer window, so a small ring is enough — and it stops a long session
 * accumulating one entry per move forever. */
const SETTLED_TOKENS_MAX = 64;

export const EVENT_TRANSFER_SETTLED = "transfer:settled";

export interface AdoptionPayload {
  readonly paneId: number;
  readonly cwd: string | null;
  readonly agentId: string | null;
  readonly scrollback: string;
  readonly cols: number;
  readonly rows: number;
  readonly tabName: string | null;
  readonly dotColor: string | null;
  readonly workspacePath: string | null;
}

export type AbortReason =
  /** A window called abort — the destination refused, or the source changed
   * its mind. */
  | "requested"
  /** No commit within TRANSFER_TIMEOUT_MS. */
  | "timedOut"
  /** The held-back output passed BUFFER_MAX_BYTES. */
  | "bufferFull"
  /** A window the transfer depends on was destroyed or is closing. */
  | "windowGone";

type Settled =
  | { readonly outcome: "committed" }
  | { readonly outcome: "aborted"; readonly reason: AbortReason };

interface BufferedEvent {
  readonly event: string;
  readonly payload: unknown;
}

interface Transfer {
  readonly from: string;
  /** Null until `claim`. */
  to: string | null;
  /** Window a pending adoption was registered for, before it claims. Lets a
   * destination that dies before `claim` still abort the transfer. */
  reservedTo: string | null;
  readonly token: string;
  staged: AdoptionPayload | null;
  buffered: BufferedEvent[];
  bufferedBytes: number;
  /** The PTY exited mid-transfer. The route entry must outlive that so the
   * buffered exit event still reaches the destination. */
  exited: boolean;
  readonly started: number;
}

type PaneRoute =
  | { readonly kind: "owned"; readonly label: string }
  | { readonly kind: "transferring"; readonly transfer: Transfer };

/** Where the coordinator emits. Production sends to a window; tests record,
 * which is the only way to assert delivery ORDER. */
export type EventSink = (
  label: string,
  event: string,
  payload: unknown,
) => void;

export type PaneAccessErrorKind =
  "not-routed" | "transferring" | "owned-by-other";

export class PaneAccessError extends Error {
  constructor(
    readonly kind: PaneAccessErrorKind,
    readonly paneId: number,
    readonly owner?: string,
  ) {
    super(
      kind === "not-routed"
        ? `Pane #${paneId} is not registered`
        : kind === "transferring"
          ? `Pane #${paneId} is being moved to another window`
          : `Pane #${paneId} belongs to window ${owner}`,
    );
    this.name = "PaneAccessError";
  }
}

export class WindowCoordinator {
  private readonly routes = new Map<number, PaneRoute>();
  private readonly settled = new Map<string, Settled>();
  private readonly settledOrder: string[] = [];
  /** Labels of destroyed windows. Never reused within a process run, so this
   * cannot outgrow the number of windows opened. */
  private readonly dead = new Set<string>();
  /** Panes a settle handed back to a window that no longer exists; the caller
   * drains and kills them. */
  private pendingOrphans: number[] = [];
  private nextToken = 0;

  constructor(private readonly sink: EventSink) {}

  register(paneId: number, windowLabel: string): void {
    this.routes.set(paneId, { kind: "owned", label: windowLabel });
  }

  unregister(paneId: number): void {
    const route = this.routes.get(paneId);
    if (route?.kind === "transferring") {
      // Mid-transfer the entry must outlive the PTY: it holds the exit event
      // the destination is owed on commit. `settle` drops the route instead of
      // re-owning it.
      route.transfer.exited = true;
      return;
    }
    this.routes.delete(paneId);
  }

  /** The owning window, or null while a transfer is open — a transferring pane
   * has no owner, and saying otherwise would let a caller act on it. */
  owner(paneId: number): string | null {
    const route = this.routes.get(paneId);
    return route?.kind === "owned" ? route.label : null;
  }

  /** Pane ids still owned by this window, for close-window dispose. */
  panesForWindow(windowLabel: string): number[] {
    const ids: number[] = [];
    for (const [id, route] of this.routes) {
      if (route.kind === "owned" && route.label === windowLabel) {
        ids.push(id);
      }
    }
    return ids;
  }

  /**
   * Every live pane, transferring ones included.
   *
   * Deliberately disagrees with `panesForWindow`: that answers "what do I kill
   * when this window closes", where a mid-transfer pane must be left alone;
   * this answers "is anything busy", where missing a mid-transfer pane kills a
   * running agent without a prompt.
   */
  allPanes(): number[] {
    return [...this.routes.keys()];
  }

  /** Route one PTY event. Emission is synchronous and in-order, so a chunk
   * read during a commit cannot overtake the flush that commit performs. */
  deliver(
    paneId: number,
    event: string,
    payload: unknown,
    now: number = Date.now(),
  ): void {
    this.sweep(now);
    const route = this.routes.get(paneId);
    if (route === undefined) {
      // No broadcast fallback — see the header.
      console.warn(`Deck: no route for pane ${paneId}, dropping ${event}`);
      return;
    }
    if (route.kind === "owned") {
      this.sink(route.label, event, payload);
      return;
    }
    const { transfer } = route;
    transfer.bufferedBytes += estimateBytes(payload);
    transfer.buffered.push({ event, payload });
    if (transfer.bufferedBytes > BUFFER_MAX_BYTES) {
      console.warn(
        `Deck: transfer buffer for pane ${paneId} passed ${BUFFER_MAX_BYTES} bytes, returning it to window ${transfer.from}`,
      );
      // The overflowing chunk was already pushed, so the flush carries it too.
      this.settle(paneId, transfer.from, {
        outcome: "aborted",
        reason: "bufferFull",
      });
    }
  }

  /**
   * Abandon every transfer that outlived the timeout, returning each pane to
   * its source. Called by the PTY commands as well as the transfer commands:
   * mid-transfer writes are rejected, so a pane whose destination died produces
   * no output and would otherwise never be swept.
   */
  sweep(now: number = Date.now()): void {
    const expired: Array<[number, string]> = [];
    for (const [id, route] of this.routes) {
      if (
        route.kind === "transferring" &&
        now - route.transfer.started >= TRANSFER_TIMEOUT_MS
      ) {
        expired.push([id, route.transfer.from]);
      }
    }
    for (const [paneId, source] of expired) {
      console.warn(
        `Deck: transfer for pane ${paneId} timed out, returning it to window ${source}`,
      );
      this.settle(paneId, source, { outcome: "aborted", reason: "timedOut" });
    }
  }

  /** Open a transfer for a pane this window owns. Output starts buffering the
   * moment this returns. */
  beginTransfer(
    from: string,
    paneId: number,
    now: number = Date.now(),
  ): string {
    // Before the route check, so a pane whose previous transfer expired can
    // start a new one.
    this.sweep(now);
    const route = this.routes.get(paneId);
    if (route === undefined) {
      throw new Error(`Pane #${paneId} is not registered`);
    }
    if (route.kind === "transferring") {
      throw new Error(`Pane #${paneId} is already being transferred`);
    }
    if (route.label !== from) {
      throw new Error(`Pane #${paneId} is owned by window ${route.label}`);
    }
    this.nextToken += 1;
    const token = `xfer-${this.nextToken}`;
    this.routes.set(paneId, {
      kind: "transferring",
      transfer: {
        from,
        to: null,
        reservedTo: null,
        token,
        staged: null,
        buffered: [],
        bufferedBytes: 0,
        exited: false,
        started: now,
      },
    });
    return token;
  }

  /**
   * The source puts up the adoption payload it serialized after `prepare`
   * quiesced the stream. Separate from `prepare` because the payload does not
   * exist yet when `prepare` returns — without this the source has no route to
   * hand its serialized buffer to `claim`.
   */
  stagePayload(
    token: string,
    caller: string,
    payload: AdoptionPayload,
    now: number = Date.now(),
  ): void {
    this.sweep(now);
    const found = this.findTransfer(token);
    if (found === null) {
      throw new Error(`Transfer ${token} is not open`);
    }
    const { transfer } = found;
    if (transfer.from !== caller) {
      throw new Error(
        `Transfer ${token} can only be staged by window ${transfer.from}`,
      );
    }
    if (transfer.staged !== null) {
      throw new Error(`Transfer ${token} already carries a payload`);
    }
    transfer.staged = payload;
  }

  /** The destination takes the payload and records itself as the receiver. */
  claim(
    token: string,
    caller: string,
    now: number = Date.now(),
  ): AdoptionPayload {
    this.sweep(now);
    const found = this.findTransfer(token);
    if (found === null) {
      throw new Error(`Transfer ${token} is not open`);
    }
    const { transfer } = found;
    if (transfer.to !== null) {
      throw new Error(`Transfer ${token} was already claimed`);
    }
    if (transfer.staged === null) {
      throw new Error(`Transfer ${token} has no staged payload`);
    }
    transfer.to = caller;
    return transfer.staged;
  }

  /** Name the window a pending adoption was opened for, before it claims, so a
   * destination dying before `claim` still aborts the transfer. */
  reserveDestination(token: string, label: string): void {
    const found = this.findTransfer(token);
    if (found === null) {
      throw new Error(`Transfer ${token} is not open`);
    }
    found.transfer.reservedTo = label;
  }

  /** Hand the pane to the window that claimed it, flushing what buffered. */
  commit(token: string, caller: string, now: number = Date.now()): void {
    this.sweep(now);
    const settled = this.settled.get(token);
    if (settled !== undefined) {
      // Idempotent by token: a retry after the fact still gets an answer.
      if (settled.outcome === "committed") {
        return;
      }
      throw new Error(`Transfer ${token} was aborted`);
    }
    const found = this.findTransfer(token);
    if (found === null) {
      throw new Error(`Transfer ${token} is not open`);
    }
    if (found.transfer.to !== caller) {
      throw new Error(
        `Transfer ${token} can only be committed by the window that claimed it`,
      );
    }
    this.settle(found.paneId, caller, { outcome: "committed" });
  }

  /**
   * Return the pane to its source, flushing what buffered.
   *
   * Any caller may abort: abort never moves a pane anywhere it was not
   * already, so there is nothing to guard, and a destination that failed
   * before it claimed still needs to release the pane.
   */
  abort(token: string, now: number = Date.now()): void {
    this.sweep(now);
    const settled = this.settled.get(token);
    if (settled !== undefined) {
      if (settled.outcome === "aborted") {
        return;
      }
      throw new Error(`Transfer ${token} was already committed`);
    }
    const found = this.findTransfer(token);
    if (found === null) {
      throw new Error(`Transfer ${token} is not open`);
    }
    this.settle(found.paneId, found.transfer.from, {
      outcome: "aborted",
      reason: "requested",
    });
  }

  /**
   * Abort every transfer this window takes part in, in either role. Called on
   * close BEFORE the busy guard: a transfer left open across a close would
   * hold the pane frozen until the timeout, and the guard would then run
   * against a route nobody owns.
   */
  abortInvolving(label: string, now: number = Date.now()): void {
    this.sweep(now);
    const involved: Array<[number, string]> = [];
    for (const [id, route] of this.routes) {
      if (route.kind !== "transferring") {
        continue;
      }
      const { transfer } = route;
      if (
        transfer.from === label ||
        transfer.to === label ||
        transfer.reservedTo === label
      ) {
        involved.push([id, transfer.from]);
      }
    }
    for (const [paneId, source] of involved) {
      this.settle(paneId, source, {
        outcome: "aborted",
        reason: "windowGone",
      });
    }
  }

  /** May `caller` act on this pane? Mid-transfer the answer is no for
   * everyone, including the source. */
  assertAccess(paneId: number, caller: string): void {
    const route = this.routes.get(paneId);
    if (route === undefined) {
      throw new PaneAccessError("not-routed", paneId);
    }
    if (route.kind === "transferring") {
      throw new PaneAccessError("transferring", paneId);
    }
    if (route.label !== caller) {
      throw new PaneAccessError("owned-by-other", paneId, route.label);
    }
  }

  /** Take the panes that settled onto a dead window; the caller kills them.
   * `sweep` and `abort` can strand a pane this way and, unlike
   * `handleWindowDestroyed`, they have no orphan pass of their own. */
  takePendingOrphans(): number[] {
    const orphans = this.pendingOrphans;
    this.pendingOrphans = [];
    return orphans;
  }

  /**
   * Apply the window-death transition table and return the panes nothing will
   * otherwise kill.
   *
   * The rules are NOT symmetric: a dead destination aborts the transfer, a
   * dead source does not — the destination can still claim and commit.
   */
  handleWindowDestroyed(label: string, now: number = Date.now()): number[] {
    this.sweep(now);
    this.dead.add(label);

    // Transfers this window was going to receive — claimed, or merely reserved
    // by a window that died before it could claim.
    const aborting: Array<[number, string]> = [];
    for (const [id, route] of this.routes) {
      if (route.kind !== "transferring") {
        continue;
      }
      const { transfer } = route;
      if (transfer.to === label || transfer.reservedTo === label) {
        aborting.push([id, transfer.from]);
      }
    }
    for (const [paneId, source] of aborting) {
      this.settle(paneId, source, {
        outcome: "aborted",
        reason: "windowGone",
      });
    }

    // Orphans: panes owned by a window that no longer exists. This is the
    // crash path — no close request fired and no busy guard ran, so nothing
    // else will ever kill them.
    const orphans: number[] = [];
    for (const [id, route] of this.routes) {
      if (route.kind === "owned" && this.dead.has(route.label)) {
        orphans.push(id);
      }
    }
    for (const id of orphans) {
      this.routes.delete(id);
    }
    // Anything an earlier sweep or abort stranded on a dead window, plus what
    // the aborts above just stranded — they never entered `routes`, so the
    // pass above cannot see them.
    orphans.push(...this.takePendingOrphans());
    return orphans;
  }

  private findTransfer(
    token: string,
  ): { paneId: number; transfer: Transfer } | null {
    for (const [paneId, route] of this.routes) {
      if (route.kind === "transferring" && route.transfer.token === token) {
        return { paneId, transfer: route.transfer };
      }
    }
    return null;
  }

  /**
   * Close a transfer: flush every buffered event to `label` in append order,
   * then hand the route over.
   */
  private settle(paneId: number, label: string, outcome: Settled): void {
    const route = this.routes.get(paneId);
    if (route === undefined || route.kind !== "transferring") {
      return;
    }
    const { transfer } = route;
    this.routes.delete(paneId);

    for (const buffered of transfer.buffered) {
      this.sink(label, buffered.event, buffered.payload);
    }
    this.announceSettled(transfer, outcome);

    if (transfer.exited) {
      // The PTY already exited and deferred its unregister so the buffered
      // exit above could still be delivered. Honour it now rather than writing
      // an owned route for a pane that no longer exists.
    } else if (this.dead.has(label)) {
      // A pane may only be handed to a window that still exists. Otherwise the
      // route would name a dead label and every later chunk would be dropped
      // for the rest of the process run — queue it for the kill instead.
      this.pendingOrphans.push(paneId);
    } else {
      this.routes.set(paneId, { kind: "owned", label });
    }
    this.rememberSettled(transfer.token, outcome);
  }

  /**
   * Tell both ends how the transfer ended — AFTER the flush, so a destination
   * has every buffered byte before it learns the pane is its own.
   *
   * A label whose window is already gone still gets the emit; the sink throws
   * it away, exactly as for any other event aimed at a dead window.
   */
  private announceSettled(transfer: Transfer, outcome: Settled): void {
    const payload = { token: transfer.token, ...outcome };
    this.sink(transfer.from, EVENT_TRANSFER_SETTLED, payload);
    // Whoever claimed, or — when nobody did — whoever a window was opened for.
    // A boot-adopt window that died before claiming must learn the transfer is
    // over rather than wait out the timeout.
    const other = transfer.to ?? transfer.reservedTo;
    if (other !== null && other !== transfer.from) {
      this.sink(other, EVENT_TRANSFER_SETTLED, payload);
    }
  }

  private rememberSettled(token: string, outcome: Settled): void {
    if (!this.settled.has(token)) {
      this.settledOrder.push(token);
    }
    this.settled.set(token, outcome);
    while (this.settledOrder.length > SETTLED_TOKENS_MAX) {
      const oldest = this.settledOrder.shift();
      if (oldest !== undefined) {
        this.settled.delete(oldest);
      }
    }
  }
}

/** Serialized size of one buffered event. Only called while a transfer is
 * open, i.e. for tens of milliseconds per move. */
function estimateBytes(payload: unknown): number {
  try {
    return Buffer.byteLength(JSON.stringify(payload) ?? "");
  } catch {
    return 0;
  }
}
